frame/support/src/storage/migration.rs

```diff
@@ -24,128 +24,56 @@ use crate::hash::ReversibleStorageHasher;
+	let mut prefix = Vec::new();
+	prefix.extend_from_slice(&Twox128::hash(module));
+	prefix.extend_from_slice(&Twox128::hash(item));
+	prefix.extend_from_slice(suffix);
+	let previous_key = prefix.clone();
+	let closure = |raw_key_without_prefix: &[u8], raw_value: &[u8]| {
+		let value = T::decode(&mut &raw_value[..])?;
+		Ok((raw_key_without_prefix.to_vec(), value))
+	};
 
-/// Utility to iterate through raw items in storage.
-pub struct StorageKeyIterator<K, T, H: ReversibleStorageHasher> {
-	prefix: Vec<u8>,
-	previous_key: Vec<u8>,
-	drain: bool,
-	_phantom: ::sp_std::marker::PhantomData<(K, T, H)>,
+	PrefixIterator { prefix, previous_key, drain: false, closure }
 }
 
-impl<K, T, H: ReversibleStorageHasher> StorageKeyIterator<K, T, H> {
-	/// Construct iterator to iterate over map items in `module` for the map called `item`.
-	pub fn new(module: &[u8], item: &[u8]) -> Self {
-		Self::with_suffix(module, item, &[][..])
-	}
-
-	/// Construct iterator to iterate over map items in `module` for the map called `item`.
-	pub fn with_suffix(module: &[u8], item: &[u8], suffix: &[u8]) -> Self {
-		let mut prefix = Vec::new();
-		prefix.extend_from_slice(&Twox128::hash(module));
-		prefix.extend_from_slice(&Twox128::hash(item));
-		prefix.extend_from_slice(suffix);
-		let previous_key = prefix.clone();
-		Self { prefix, previous_key, drain: false, _phantom: Default::default() }
-	}
-
-	/// Mutate this iterator into a draining iterator; items iterated are removed from storage.
-	pub fn drain(mut self) -> Self {
-		self.drain = true;
-		self
-	}
+/// Construct iterator to iterate over map items in `module` for the map called `item`.
+pub fn storage_key_iter<K: Decode + Sized, T: Decode + Sized, H: ReversibleStorageHasher>(
+	module: &[u8],
+	item: &[u8],
+) -> PrefixIterator<(K, T)> {
+	storage_key_iter_with_suffix::<K, T, H>(module, item, &[][..])
 }
 
-impl<K: Decode + Sized, T: Decode + Sized, H: ReversibleStorageHasher> Iterator
-	for StorageKeyIterator<K, T, H>
-{
-	type Item = (K, T);
-
-	fn next(&mut self) -> Option<(K, T)> {
-		loop {
-			let maybe_next = sp_io::storage::next_key(&self.previous_key)
-				.filter(|n| n.starts_with(&self.prefix));
-			break match maybe_next {
-				Some(next) => {
-					self.previous_key = next.clone();
-					let mut key_material = H::reverse(&next[self.prefix.len()..]);
-					match K::decode(&mut key_material) {
-						Ok(key) => {
-							let maybe_value = frame_support::storage::unhashed::get::<T>(&next);
-							match maybe_value {
-								Some(value) => {
-									if self.drain {
-										frame_support::storage::unhashed::kill(&next);
-									}
-									Some((key, value))
-								}
-								None => continue,
-							}
-						}
-						Err(_) => continue,
-					}
-				}
-				None => None,
-			}
-		}
-	}
+/// Construct iterator to iterate over map items in `module` for the map called `item`.
+pub fn storage_key_iter_with_suffix<K: Decode + Sized, T: Decode + Sized, H: ReversibleStorageHasher>(
```

Comment:
It might also be good to move these standalone functions to be methods of `PrefixIterator`. Let me know what you think.

Reply:
if we deprecate the previous type, maybe we can move these function into PrefixIterator, and we could also reexport Prefixiterator in the migration module.

But standalone functions are ok to me

Reply:
Maybe we should deprecate these types instead of removing them.

maybe we can move these function into PrefixIterator, and we could also reexport Prefixiterator in the migration module.

But standalone functions are ok to me